File: DAA-LAB/DFS.cpp
#include <stdio.h>

int graph[10][10] = {
   {0, 1, 1, 0, 0},
   {1, 0, 1, 1, 0},
   {1, 1, 0, 0, 1},
   {0, 1, 0, 0, 1},
   {0, 0, 1, 1, 0}
};

int visited[10];
int n = 5;

/* Function to perform DFS traversal */
void dfs(int vertex) {
   int i;
   visited[vertex] = 1;
   printf("%d ", vertex + 1);

   for(i = 0; i < n; i++) {
      if(graph[vertex][i] == 1 && visited[i] == 0) {
         dfs(i);
      }
   }
}

/* Main function */
int main() {
   int i;
   
   for(i = 0; i < n; i++) {
      visited[i] = 0;
   }
   
   printf("DFS Traversal starting from vertex 1: ");
   dfs(0);
   
   printf("\n");
   
   return 0;
}



/*
Algorithm: Depth-First Search (DFS) using Adjacency Matrix (Recursive)

1. Initialize a visited array of size equal to the number of vertices and set all entries to 0.

2. Define a recursive DFS function:
   a. If the current node is already visited, return (base case).
   b. Print the current node (i.e., process it).
   c. Mark the current node as visited.

3. For all vertices in the graph:
   a. If the current node is connected to vertex `i` (graph[node][i] == 1)
      and vertex `i` has not been visited:
      - Recursively call DFS on vertex `i`.

4. In the main function:
   a. Define and initialize the adjacency matrix representing the graph.
   b. Initialize all entries in the visited array to 0.
   c. Call the DFS function with the starting node.

5. The order of printed nodes is the DFS traversal order.
*/

File: AI/dfs.py
# DFS
# 10 / 03 / 2025
# DFS

graph = {
    '1' : ['2', '3', '4'],
    '2' : ['5', '6'],
    '3' : [],
    '4' : ['7', '8'],
    '5' : ['9', '10'],
    '6' : [],
    '7' : [],
    '8' : [],
    '9' : [],
    '10' : []
}

visited = set()

def dfs(visited, graph, node):
  if node not in visited:
    print(node)
    visited.add(node)
    for neighbour in graph[node]:
      dfs(visited, graph, neighbour)

print("Following is the Depth-First Search")
dfs(visited, graph, '1')



# Algorithm for Depth-First Search (DFS)

# Step 1: Start

# Step 2: Represent the graph using an adjacency list (dictionary format)

# Step 3: Create an empty set named 'visited' to keep track of visited nodes

# Step 4: Define a DFS function that takes the visited set, graph, and current node as input

# Step 5: Inside the DFS function:
#         - If the current node is not in the visited set:
#             a. Print or process the node
#             b. Add the node to the visited set
#             c. For each neighbor of the current node:
#                 - Recursively call the DFS function on the neighbor

# Step 6: Call the DFS function with the starting node

# Step 7: End
